User login and devices list

wpearse opened the pull request:
This PR contains the following changes:
- Added a server input field on login screen (looked like there was the makings of a settings button, but I needed to be able to test login and didn't want to go to effort of settings screen)
- Added a basic User model (singleton, not sure how you feel about that) with properties that mimic the API response
- Added a Traccar class with a static function that tries to auth against the API
- Altered the map view so it centers the map on the user's location once they're logged-in
- Added a Logout button to the map view
- Shift the controls on the login screen up a bit so they wouldn't be obscured by the keyboard
- Added CocoaPods for dependency management
- Implemented text field delegates so you can move between inputs on the login view using only the keyboard
- Added Traccar logo to launch screen (used to just be plain white screen)

What's missing:

~~\- Set up Alamofire to use cookies, so the user info will actually be stored~~
~~\- Change the User logout() function to nuke the cookie jar (and then fix User isAuthenticated)~~

wpearse committed:
Add logo to launch screen

wpearse committed:
Ignore personal Xcode data

wpearse committed:
Use CocoaPods for dependencies

wpearse committed:
Add User model, Traccar login (http only?), center map on login

wpearse commented:
![simulator screen shot 4 05 2016 7 24 32 pm](https://cloud.githubusercontent.com/assets/2335460/15007683/ede59c3a-122e-11e6-86a9-b693d79f78e1.png)
![simulator screen shot 4 05 2016 7 24 04 pm](https://cloud.githubusercontent.com/assets/2335460/15007685/ee280912-122e-11e6-8c54-d7f41556b640.png)

wpearse committed:
Handle invalid email/password

wpearse committed:
Auto-focus on email field

tananaev commented:
First of all, can you please check Traccar Manager for reference. It should looks as similar as possible. For example, check how server address is set up on Android.

Also, can you please use Carthage for dependencies instead of CocoaPods.

Some more minor comments:

1. `Traccar` is not a good name for a class. I would recommend to follow Android naming pattern to make it more consistent.
2. Why does `Traccar` class extend `NSObject`?
3. Schema should be part of the configurable URL as some people might want to use SSL.

wpearse committed:
Rename Traccar -> WebService, add support for URL scheme

wpearse commented:
Can you please attach screenshot(s) of the relevant screens in the Android app? I do not have an Android device, and the screenshots on the Google Play store do not show the settings options available.

I have never used Carthage, but I am happy to do so if you wish.

1. Renamed to WebService.
2. Habit, removed.
3. OK, done.

wpearse committed:
Use Carthage instead of CocoaPods

wpearse committed:
Remember cookies from API

tananaev commented:
I think Alamofire manages cookies for you by default. There is no need to set it up explicitly.

tananaev commented:
By the way, you have a comment 
> the api still returns a 200 status code (and empty response) on failed login...
This is not right. If it's the case, I need to fix it on the server side. Don't need to handle this situation on the client.

tananaev commented:
When/if you start implementing websocket connection, I would recommend to use `SocketRocket`. That's what I was going to use. That's the only library I found that allows to send cookies header. Here is a code snippet:
```swift
        socket = SRWebSocket(URL: NSURL(string: ViewController.serviceUrl + "/socket"))
        socket?.requestCookies = NSHTTPCookieStorage.sharedHTTPCookieStorage().cookiesForURL(NSURL(string: ViewController.serviceUrl)!)
        socket?.delegate = self
```

wpearse committed:
Check status code instead of empty response for failed login

wpearse commented:
Of course! I had a brain fart. We now check the response code for the login success/failure.

I'll need to check up on the Alamofire cookie handling.

SocketRocket looks nice.

tananaev commented:
Here is Android login screenshot:

![screenshot_20160504-212453](https://cloud.githubusercontent.com/assets/1161863/15010148/b14dd2dc-123e-11e6-8242-3ff5d3693731.png)

When you click on the cog button at the top, there is a dialog box to edit server URL.

tananaev commented:
The way it works on Android at the moment, it actually saves user credentials in the preferences (analogue of NSUserDefaults). Then when you run the app second time, it goes straight to the map.

wpearse committed:
No need to build for platforms other than iOS

wpearse committed:
Fetch and parse devices from API

wpearse committed:
Anton's right. No need to explicitly set up cookie management with Alamofire

wpearse commented:
You were correct about the cookie handling. I've taken out that unnecessary code from the app delegate.

tananaev commented:
Let me know when it's finished for final review.

wpearse committed:
Parse device update date from string

wpearse committed:
Show list of devices

wpearse committed:
Save server and email

wpearse committed:
Disable compiler optimisations for weird EXEC BAD ACCESS

wpearse committed:
Disable devices button until data loaded

wpearse commented:
I added a few more bits and pieces:

- Remember server url and user email address
- Show devices list (only: name, last update time)

It's all finished for final review now. Sorry for opening the PR a bit early :-/

tananaev commented:
I haven't had a chance to run it yet, but the code looks good. I have only one comment:

Why do you have two `authenticate` methods? The body seems almost the same.

wpearse committed:
Remove static func (use WebService.sharedInstance.authenticate instead)

wpearse commented:
The static function has now been removed. I think I had too many beers.

tananaev commented:
Looks good as a draft. There are some minor issues, but we can fix it later. I've merged the request. Thanks.

Next step is to implement WebSocket connection and get devices to show up on the map. After that I can publish the app to the App Store.

wpearse commented:
Cool. I got devices on the map view, and basic device info done this morning. I'll look at the web socket stuff next.

tananaev commented:
Great. If you have any questions on how the API works, feel free to ask.

wpearse commented:
Hrm, ok. I'm using this code to open the web socket:

```
private func enableWebSocket() {
        socket = SRWebSocket(URL: NSURL(string: serverURL! + "socket")) // http://<my server>/socket
        if let s = socket {
            s.requestCookies = NSHTTPCookieStorage.sharedHTTPCookieStorage().cookiesForURL(NSURL(string: serverURL!)!)
            s.delegate = self
            s.open()
        }
    }
```

And I'm getting this response:

```
{NSLocalizedDescription=received bad response code from server 404, HTTPResponseStatusCode=404}
```

Any ideas?

wpearse commented:
If I use `api/socket` I get  `received bad response code from server 503`.

tananaev commented:
You need to check response content for more details about the error.

wpearse commented:
Server response:

```
Request Version: HTTP/1.1
Status Code: 503
Response Phrase: Endpoint Creation Failed
```

tananaev commented:
Let's discuss it in the next PR.
